File: ROS2/ros2_ws/src/basics_pubsub/basics_pubsub/simple_pub.py
import rclpy
from rclpy.node import Node
from std_msgs.msg import String

class SimplePub(Node):
    def __init__(self):
        super().__init__('simple_pub')
        self.pub = self.create_publisher(String, 'chatter', 10)
        self.timer = self.create_timer(0.5, self.tick)
        self.count = 0

    def tick(self):
        msg = String()
        msg.data = f"hello {self.count}"
        self.pub.publish(msg)
        self.get_logger().info(f"sent: {msg.data}")
        self.count += 1

def main():
    rclpy.init()
    node = SimplePub()
    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        node.get_logger().info("Shutting down")
    finally:
        node.destroy_node()
        # Only call shutdown if the context is still active
        if rclpy.ok():
            rclpy.shutdown()



if __name__ == '__main__':
    main()

File: ROS2/ros2_ws/src/basics_pubsub_cpp/src/simple_pub.cpp
#include <chrono>
#include <memory>
#include "rclcpp/rclcpp.hpp"
#include "std_msgs/msg/string.hpp"


using namespace std::chrono_literals;

class SimplePub : public rclcpp::Node {
public:
  SimplePub() : Node("simple_pub_cpp"), count_(0) {
    pub_ = this->create_publisher<std_msgs::msg::String>("chatter_cpp", 10);
    timer_ = this->create_wall_timer(500ms, [this]() {
      auto msg = std_msgs::msg::String();
      msg.data = "hello " + std::to_string(count_++);
      pub_->publish(msg);
      RCLCPP_INFO(this->get_logger(), "sent: %s", msg.data.c_str());
    });
  }
private:
  rclcpp::Publisher<std_msgs::msg::String>::SharedPtr pub_;
  rclcpp::TimerBase::SharedPtr timer_;
  int count_;
};

int main(int argc, char * argv[]) {
  rclcpp::init(argc, argv);
  auto node = std::make_shared<SimplePub>();
  rclcpp::spin(node);
  rclcpp::shutdown();
  return 0;
}